### PR TITLE
Icon support

### PR DESCRIPTION
**Pull request by andreaspauling**
- Module update_phenology_realtime allows now ICON input and writes ICON output
- I/O designed to be close to opr setup at MCH
- cfgrib replaced by eccodes for I/O

### CODE REVIEW

**Review by sadamov — DISMISSED** [review]
Thanks @andreaspauling :rocket: I am glad that this now runs with ICON! 
Please make sure that the GitHub actions are successful before merging the PR.
I ran the code on Balfrin and the update_strength_realtime function is broken. (error message below) I have some more general comments since this is the initial ICON release:

- describe a bit clearer in the README how the user should interact with the main.py script and the notebooks for plotting. 
- some notebooks for plotting are in a broken state, I would either remove or fix them. imports seem to be missing
- consider adding logging/debug levels. the code is quite verbose 

```
/users/sadamov/pyprojects/realtime_pollen_calibration/src/realtime_pollen_calibration/utils.py:108: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data = pd.read_csv(
/users/sadamov/pyprojects/realtime_pollen_calibration/src/realtime_pollen_calibration/utils.py:124: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data_mod = pd.read_csv(
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/users/sadamov/pyprojects/realtime_pollen_calibration/src/realtime_pollen_calibration/update_strength_realtime.py", line 53, in update_strength_realtime
    utils.to_grib(file_in, file_out, dict_fields)
  File "/users/sadamov/pyprojects/realtime_pollen_calibration/src/realtime_pollen_calibration/utils.py", line 561, in to_grib
    dict_fields[short_name][values == 0] = 0
IndexError: boolean index did not match indexed array along dimension 0; dimension is 786 but corresponding boolean dimension is 919620
>>> 
```

File: data/grib2_files_ICON-CH1/ART_POV_iconR19B08-grid_0001_all_specs_values
Comment (andreaspauling) [reply data/grib2_files_ICON-CH1/ART_POV_iconR19B08-grid_0001_all_specs_values:1]: The test data is not part of the repo any more. It will be available from a location defined by osm.

File: data/grib2_files_ICON-CH1/POV_out
Comment (andreaspauling) [reply data/grib2_files_ICON-CH1/POV_out:1]: dito

File: src/realtime_pollen_calibration/update_phenology_realtime.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/update_phenology_realtime.py]: Done.

File: src/realtime_pollen_calibration/update_phenology_realtime.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/update_phenology_realtime.py]: Done. Now there is a config.yaml.

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:33]: These values should not be changed by users. Hence I would leave them there

**Comment by andreaspauling** [comment]
@clairemerker Die Pollenkalibration wär jetzt parat für deinen Review. Es ist so ziemlich alles umgestellt... DaniL wird das auch anschauen, da osm das dann betreiben darf ;-) Trotzdem wäre es vielleicht gut, wenn du noch einen Blick darüber werfen könnest

**Comment by DanielLeuenberger** [comment]
Thanks for the README. It contains useful information for the configuration of the package and how to run it. Please add the following information to the README:

General questions:
- What happens, if there are no(t enough) pollen fields in the POV_infile, the station_obs_file, station_mod_file?
- How do we initialize a season? Then, we do not have 120h of past pollen from the model history. Can fieldextra fill all 120h with missing values?


1. Section "How to configure the package"

- Specify in more detail the fields that need to be present in the different grib and atab files: variable(s) (I have nothing found "specified above"), date and time, restrictions to grib2 metadata (such as process generating ID etc) if any.
- "last 120 hours" relative to which date/time? To that of the fields in POV_infile? Please document this.
- what is the hour_incr for? Are there use cases in which it is need to be set to different values than 1?

3. Section "How to run the package"

- I do not understand the paragraph "The phenological model of ICON..." Please document in more details what is exactly needed to run and updated at which time of the day. Does the calibration needs to be run before the respective KENDA-CH1 cycle? We can enforce that with dependencies in the LMPackage
- what are the "phenological fields"? Please specify in detail.
- Please make sure that the script does not need to be in the root directory of the installed package and that the script can be run from anywhere!

File: src/realtime_pollen_calibration/update_strength.py
Comment (clairemerker) [inline src/realtime_pollen_calibration/update_strength.py:26]: There is also a function `read_pov_file` in the file `update_phenology.py`. It is hard to tell why they are different without a doc string. Maybe consider adding one to justify the duplication.

File: README.md
Comment (clairemerker) [inline README.md:60]: libmamba is the default server for conda versions > 23.10. You could consider removing mamba from the setup script to make it leaner in a future PR .

**Review by clairemerker — COMMENTED** [review]
Nice work! A lot of refactoring and added documentation! 
The scope of the PR was quite big because it mixes code changes and the blueprint update, so I didn't manage to look at everything in details. But it looks good to me overall! 

I added a few comments to the code, and I have two comments regarding testing: 
- I don't have read access to the test data so I couldn't copy it to test the package. I understood that the test data will be moved in a centralised place, so this issue will probably be solved then. 
- Consider adding at least integration tests for the two tools `update_phenology` and `update_strength`. I run the commands described in the readme and they worked, which is very nice but I didn't expect it since it was unclear to me if the config file was setup in a way that would run for all users. Also, I have no possibility to check if the output is correct. It would be great if this could be integrated as a test, since everything needed for the test seems to already be present :)

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:33]: Done

File: src/realtime_pollen_calibration/mutable_number.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/mutable_number.py:1]: It is  just a left-over from the template example and is removed now.

File: README.md
Comment (andreaspauling) [reply README.md:60]: Done.

File: README.md
Comment (andreaspauling) [reply README.md:115]: The values above are now documented in utils.py.
The fields in that section are now exlicitly listed.

File: src/realtime_pollen_calibration/update_strength.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/update_strength.py:26]: The two functions are not doing the same. They could be unified, to be discussed. For now I added a docstring.

**Comment by andreaspauling** [comment]
> Thanks for the README. It contains useful information for the configuration of the package and how to run it. Please add the following information to the README:
> 
> General questions:
> 
> * What happens, if there are no(t enough) pollen fields in the POV_infile, the station_obs_file, station_mod_file?
> * How do we initialize a season? Then, we do not have 120h of past pollen from the model history. Can fieldextra fill all 120h with missing values?
> 
> Section "How to configure the package"
> 
> * Specify in more detail the fields that need to be present in the different grib and atab files: variable(s) (I have nothing found "specified above"), date and time, restrictions to grib2 metadata (such as process generating ID etc) if any.
> * "last 120 hours" relative to which date/time? To that of the fields in POV_infile? Please document this.
> * what is the hour_incr for? Are there use cases in which it is need to be set to different values than 1?
> 
> Section "How to run the package"
> 
> * I do not understand the paragraph "The phenological model of ICON..." Please document in more details what is exactly needed to run and updated at which time of the day. Does the calibration needs to be run before the respective KENDA-CH1 cycle? We can enforce that with dependencies in the LMPackage
> * what are the "phenological fields"? Please specify in detail.
> * Please make sure that the script does not need to be in the root directory of the installed package and that the script can be run from anywhere!

Replies:
General questions:
-  If at least one of the mandatory fields is missing the package exits with status 1 and tells the user.
- This is handled by the oper. workflow 

Section "How to configure the package"
- This is now documented in more detail.
- dito
- This parameter should be adapted by the user if the calibration is done for a subsequent run more than one hour ahead. This is also documented now

Section "How to run the package"
- The paragraph has been re-written. The oper. workflow is already in place.
- dito
- done

**Review by sadamov — COMMENTED** [review]
I did not review this again, I understand that other reviewers have taken over in the meantime @andreaspauling. As my previous requested changes are blocking this merge, I suggest that an admin @cosunae removes [my previous review ](https://github.com/MeteoSwiss/realtime_pollen_calibration/pull/11#pullrequestreview-1778775641)from the conversation. I'd rather not approve changes I didn't actually review and currently don't have time for it.

image

**Review by DanielLeuenberger — APPROVED** [review]
Sieht soweit gut aus für mich.

**Pull request merged**
